src/codegen/mod.rs

```diff
@@ -0,0 +1,2000 @@
+    }
+
+    fn seen(&self, item: ItemId) -> bool {
+        self.items_seen.contains(&item)
+    }
+
+    fn set_seen(&mut self, item: ItemId) {
+        self.items_seen.insert(item);
+    }
+
+    fn seen_function(&self, name: &String) -> bool {
+        self.functions_seen.contains(name)
+    }
+
+    fn saw_function(&mut self, name: &str) {
+        self.functions_seen.insert(name.into());
+    }
+
+    fn inner<F>(&mut self, cb: F) -> Vec<P<ast::Item>>
+        where F: FnOnce(&mut Self)
+    {
+        let mut new = Self::new();
+
+        cb(&mut new);
+
+        self.saw_union |= new.saw_union;
+
+        new.items
+    }
+}
+
+impl ops::Deref for CodegenResult {
+    type Target = Vec<P<ast::Item>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.items
+    }
+}
+
+impl ops::DerefMut for CodegenResult {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.items
+    }
+}
+
+struct ForeignModBuilder {
+    inner: ast::ForeignMod,
+}
+
+impl ForeignModBuilder {
+    fn new(abi: Abi) -> Self {
+        ForeignModBuilder {
+            inner: ast::ForeignMod {
+                abi: abi,
+                items: vec![],
+            }
+        }
+    }
+
+    fn with_foreign_item(mut self, item: ast::ForeignItem) -> Self {
+        self.inner.items.push(item);
+        self
+    }
+
+    #[allow(dead_code)]
+    fn with_foreign_items<I>(mut self, items: I) -> Self
+        where I: IntoIterator<Item=ast::ForeignItem>
+    {
+        self.inner.items.extend(items.into_iter());
+        self
+    }
+
+    fn build(self, ctx: &BindgenContext) -> P<ast::Item> {
+        use syntax::codemap::DUMMY_SP;
+        P(ast::Item {
+            ident: ctx.rust_ident(""),
+            id: ast::DUMMY_NODE_ID,
+            node: ast::ItemKind::ForeignMod(self.inner),
+            vis: ast::Visibility::Public,
+            attrs: vec![],
+            span: DUMMY_SP,
+        })
+    }
+}
+
+struct ArrayTyBuilder {
+    len: usize,
+}
+
+// TODO: Aster seem to lack a array builder?
+impl ArrayTyBuilder {
+    fn new() -> Self {
+        ArrayTyBuilder {
+            len: 0,
+        }
+    }
+
+    fn with_len(mut self, len: usize) -> Self {
+        self.len = len;
+        self
+    }
+
+    fn build(self, ty: P<ast::Ty>) -> P<ast::Ty> {
+        use syntax::codemap::DUMMY_SP;
+        let size =
+            ast::LitKind::Int(self.len as u64,
+                              ast::LitIntType::Unsigned(ast::UintTy::Us));
+        let size = ast::ExprKind::Lit(P(respan(DUMMY_SP, size)));
+        let array_kind = ast::TyKind::FixedLengthVec(ty,
+            P(ast::Expr {
+                id: ast::DUMMY_NODE_ID,
+                node: size,
+                span: DUMMY_SP,
+                attrs: ast::ThinVec::new(),
+            })
+        );
+
+        P(ast::Ty {
+            id: ast::DUMMY_NODE_ID,
+            node: array_kind,
+            span: DUMMY_SP,
+        })
+    }
+}
+
+struct BlobTyBuilder {
```

Comment:
It creates an appropriate type for a blob with a given `Layout`, with proper size and alignment. But I'll rather write this in a comment.